File: week07/frequency.py
#!/usr/bin/python3

import re
import sys
import glob

word = sys.argv[1].lower()
n_matches = {}
n_words = {}

for file in glob.glob("lyrics/*.txt"):
   f = open(file, "r")

   count = 0
   total = 0
   words = []

   for line in f:
      line = line.lower()
      words = re.split(r'[^a-zA-Z]+', line)
   
      for w in words:
         if w:
            total += 1
   
            if w == word:
               count += 1

   searchObj = re.search(r'lyrics\/(.*)\.txt', file)
   artist = searchObj.group(1)
   artist = re.sub(r'[^a-zA-Z]', ' ', artist)

   n_matches[artist] = count
   n_words[artist] = total   

   f.close()

for key in sorted(n_words.keys()):
   print("%4d/%6d = %.9f %s" %(n_matches[key], n_words[key], 
                               n_matches[key]/n_words[key], key))

File: week07/frequency.pl
#!/usr/bin/perl -w 

sub totalWords {
   my $num_words = 0;
   open my $F, "<", "$_[0]" or die;  
 
   while($line = <$F>) {
      undef my @words;

      $line =~ s/^[^a-zA-Z]+//;
      @words = split(/[^a-zA-Z]+/, $line);
      $num_words += @words;
   }

   close $F;
   return $num_words;
}

sub countWord {
   my $word = $_[0];
   open my $F, "<", "$_[1]" or die;

   my $count = 0;

   while($line = <$F>) {
      undef my @words;
      $line = lc $line;
   
      @words = ($line =~ /\b$word\b/g);
      $count += @words;
   }

   close $F;
   return $count;
}

$word = lc $ARGV[0];

foreach $file (glob "lyrics/*.txt") {
   $total = totalWords($file);
   $count = countWord($word, $file);  

   $file =~ s/lyrics\/(.*).txt/$1/;
   $file =~ s/[^a-zA-Z]/ /g;
   $name = $file;

   $table{$name}{"total"} = $total;
   $table{$name}{$word} = $count;
}

foreach $artist (sort keys %table) {
   printf "%4d/%6d = %.9f %s\n", $table{$artist}{$word}, $table{$artist}{"total"}, $table{$artist}{$word}/$table{$artist}{"total"}, $artist;
} 